### DISCUSSION TITLE
spring cloud sleuth and splunk otel java library

### DISCUSSION THREAD

**Discussion by hshah2811** [signalfx/splunk-otel-java#789] [Q&A]
Can anyone help me to understand how this library to be integrated with spring boot based app which is leveraging spring cloud sleuth for trace data generation?

There is experimental project spring-cloud-sleuth-otel which supports open telemetry. 

https://github.com/spring-projects-experimental/spring-cloud-sleuth-otel

Should i use  spring cloud sleuth to generate trace data or splunk-otel-java?

**Comment by mateuszrzeszutek** [comment]
Hey @hshah2811 ,
While we recommend using the javaagent (well, of course), if you decide to try Spring Sleuth you should be able to send the telemetry over to Splunk cloud as well. The javaagent has more instrumentations that Spring Sleuth, so you will probably get way more telemetry from the agent; and the telemetry data produced by Sleuth might differ slightly from the javaagent's output.
Make sure you don't use both at the same time -- the javaagent and Sleuth set up separate instances of OpenTelemetry, and will produce duplicate and/or conflicting telemetry.

**Reply by hshah2811** [reply]
Thanks @mateuszrzeszutek , 

I have removed sleuth and added agent library. 

traceId and spanIds are available in MDC context by default and I am able to log it without any code changes.

Do you know how to add custom baggage values into MDC context? I tried to access it from MDC but couldn't find it.

**Reply by laurit** [reply]
Agent only adds `trace_id`, `span_id` and `trace_flags` to MDC. If you need more values you'll have to add them to MDC yourself. If you are using log4j you could use a `ContextDataProvider`.

**Comment by hshah2811** [comment]
Hello @mateuszrzeszutek ,

Can you please help me in above question?
